kork-retrofit/src/main/java/com/netflix/spinnaker/kork/retrofit/ErrorHandlingExecutorCallAdapterFactory.java

```
@@ -0,0 +1,266 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.netflix.spinnaker.kork.retrofit;
+
+import com.netflix.spinnaker.kork.retrofit.exceptions.RetrofitException;
+import com.netflix.spinnaker.kork.retrofit.exceptions.SpinnakerHttpException;
+import com.netflix.spinnaker.kork.retrofit.exceptions.SpinnakerNetworkException;
+import com.netflix.spinnaker.kork.retrofit.exceptions.SpinnakerServerException;
+import java.io.IOException;
+import java.lang.annotation.Annotation;
+import java.lang.reflect.ParameterizedType;
+import java.lang.reflect.Type;
+import java.util.concurrent.Executor;
+import okhttp3.Request;
+import okio.Timeout;
+import org.jetbrains.annotations.NotNull;
+import org.jetbrains.annotations.Nullable;
+import org.springframework.http.HttpStatus;
+import retrofit2.Call;
+import retrofit2.CallAdapter;
+import retrofit2.Callback;
+import retrofit2.Response;
+import retrofit2.Retrofit;
+
+/*
+ * As part retrofit2 {@link retrofit.RetrofitError} and {@link retrofit.ErrorHandler} are removed.
+ * So this class helps to achieve similar logic as retrofit and handle exceptions globally in retrofit2.
+ * This can be achieved by setting this class as CallAdapterFactory at the time of {@link Retrofit} client creation.
+ *  */
+public class ErrorHandlingExecutorCallAdapterFactory extends CallAdapter.Factory {
+
+  private final @Nullable Executor callbackExecutor;
+
+  ErrorHandlingExecutorCallAdapterFactory(@Nullable Executor callbackExecutor) {
+    this.callbackExecutor = callbackExecutor;
+  }
+
+  public static ErrorHandlingExecutorCallAdapterFactory getInstance(
+      @Nullable Executor callbackExecutor) {
+    return new ErrorHandlingExecutorCallAdapterFactory(callbackExecutor);
+  }
+
+  @Nullable
+  @Override
+  public CallAdapter<?, ?> get(Type returnType, Annotation[] annotations, Retrofit retrofit) {
+
+    if (getRawType(returnType) != Call.class) {
```

This could use a comment in the code.  Why do we need this check?  Better to throw an exception if the check fails?

---

Aaah, I see [the javadoc](https://square.github.io/retrofit/2.x/retrofit/index.html?retrofit2/CallAdapter.Factory.html) says to return null if the factory can't handle it.  Still curious why we need the check.

---

I will throw SpinnakerServerException instead of returning null.

---

Returning null seems the correct behavior based on the javadoc. I'd still like a comment in the code explaining the check.